### PR TITLE
fix: install stale agent properly

### LINKED ISSUE #1643: Scenario for the upgrade agent need to be fixed
https://github.com/elastic/e2e-testing/pull/1642
This scenario:

```gherkin
@upgrade-agent
@nightly
@skip:windows
Scenario Outline: Upgrading the installed <os> agent
  Given a "<os>" agent "stale" is deployed to Fleet with "tar" installer
    And agent is in version "stale"
    And certs are installed
    And the "elastic-agent" process is "restarted" on the host
  When agent is upgraded to version "latest"
  Then agent is in version "latest"
```
this need to be fixed because is it not working as designed first stem never installed stale version.

### PR DESCRIPTION
**Pull request by mdelapenya**
- chore: add logs
- chore: add missing dependency
- chore: rename scenario for simplification
- fix: download the stale agent version properly
- revert: for testing purposes

## What does this PR do?
In this PR we are refactoring the upgrade scenario to accomplish two main things:

1. make sure the installed stale agent honours the version we are setting
1. use a matrix of stale agent versions (releases) to run the tests multiple times and possibly capture more errors

We are adding trace logs to verify the versions that are installed, and finally extracting the upgrade scenarios to its own feature file.

## Why is it important?
We detected that the scenario was always installing the current elastic-agent version for the branch as stale, which is incorrect.

Besides that, in the past, we had to update the stale version with certain frequency, when the artifacts-api retired the least minor stale version, causing noise and making the scenario to fail.

With this change, we are explicitly using released versions, which is what a user upgrading the agent would have. Instead of checking that a snapshot is upgradeable, we will rely on releases
because a snapshot could be unstable. And if we are daily bumping the target version to upgrade to, with the automated bump, then we will be detecting regressions.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have run the Unit tests (`make unit-test`), and they are passing locally
- [x] I have run the End-2-End tests for the suite I'm working on, and they are passing locally
- [ ] I have noticed new Go dependencies (run `make notice` in the proper directory)

## Related issues

- Closes #1643

### CODE REVIEW

File: e2e/_suites/fleet/features/fleet_mode_agent.feature [hunk 1]
```diff
@@ -29,14 +29,20 @@ Scenario Outline: Deploying the agent
 # | os     |
 # | debian |
 
-# @upgrade-agent
-@nightly
-Scenario Outline: Upgrading the installed agent
-  Given an agent "stale" is deployed to Fleet with "tar" installer
+@upgrade-agent
```
Comment (mdelapenya) [inline e2e/_suites/fleet/features/fleet_mode_agent.feature:32]: Will add the `@nightly` tag right after testing it

**Review by jlind23 — COMMENTED** [review]
Looks good to me but i'll let Adam review it.
Did you test it with the 8.1.1 and checked that the upgrade was failing?

**Comment by mdelapenya** [comment]
@jlind23 I can verify that upgrading using the API calls using curl, which is what the e2e basically do:

```shell
curl -X POST -u "admin:changeme" -H "kbn-xsrf:foo" -H "Content-Type: application/json" http://18.117.115.182:5601/api/fleet/agents/be9193bf-6d05-4a15-9897-4c319c13ba1a/upgrade -d {\"force\":true,\"version\":\"8.3.0-657f2d50-SNAPSHOT\"}
```

is leaving the agent in an unhealthy state therefore the tests never find the upgraded agent. OTOH, using the UI as we did past week during our pairing session, the upgrade worked.

**Comment by jlind23** [comment]
@mdelapenya the fix was pushed to the 8.1.x branch last Friday. Which tag did you use to perform the upgrade?

**Comment by mdelapenya** [comment]
> @mdelapenya the fix was pushed to the 8.1.x branch last Friday. Which tag did you use to perform the upgrade?

The upgrade-to version was `8.3.0-657f2d50-SNAPSHOT`

**Comment by jlind23** [comment]
@mdelapenya but for the 8.1.X base version?

**Comment by mdelapenya** [comment]
> @mdelapenya but for the 8.1.X base version?

The matrix is using releases, so for the release, we are using 8.1.0, 8.1.1 and 8.1.2 (see https://github.com/elastic/e2e-testing/pull/2354/files#diff-3746d5ea011663df7d08a904fbe1d56d2ed37824faaa9bbd7669a749f7657d79R14-R16)

If you mean that the upgrade-to version is 8.1.x, then we need to wait for the backports

**Review by kuisathaverat — APPROVED** [review]

**Comment by mdelapenya** [comment]
@jlind23 I've verified that using the UI causes the stale agent to be upgraded correctly, but using the tests it's not.

The code to upgrade the agent is hitting this API: `/api/fleet/agents/:agentID/upgrade` including this payload:

```json
{"version":"` + version + `"} // version to upgrade to
```

Do you know if there is anything the tests should do to cause the upgrade to finish?

**Comment by blakerouse** [comment]
I think that would be a better question for the Fleet UI team. If that API call sends the action to be sent to the Elastic Agent then it should complete the upgrade.

It is also possible to upgrade an agent without triggering the action from the Fleet UI. You can use the `elastic-agent upgrade ${version}` and it will perform the upgrade. It is even possible to have the file to upgrade to be local on the disk with `elastic-agent upgrade ${version} --source-uri file://${absolute_path_to_dir_with_files_present}`.

**Comment by mdelapenya** [comment]
> I think that would be a better question for the Fleet UI team. If that API call sends the action to be sent to the Elastic Agent then it should complete the upgrade.
> 
> It is also possible to upgrade an agent without triggering the action from the Fleet UI. You can use the `elastic-agent upgrade ${version}` and it will perform the upgrade. It is even possible to have the file to upgrade to be local on the disk with `elastic-agent upgrade ${version} --source-uri file://${absolute_path_to_dir_with_files_present}`.

I'd then let the team define what's the upgrade strategy to follow here:

1. using Kibana/Fleet APIs
2. invoking the upgrade command
3. invoking the upgrade command including an absolute path that was previously downloaded

**Comment by jlind23** [comment]
> https://github.com/elastic/e2e-testing/pull/2354#issuecomment-1097001948

@ph any thoughts on the scenario we should adopt?

**Comment by mdelapenya** [comment]
I also think that 1. fits more into the e2e space. For 2, I truly believe it falls under the integration testing space, which "should" be covered after https://github.com/elastic/elastic-agent/issues/269

**Comment by mdelapenya** [comment]
> 1. From Fleet UI (@joshdover) can help here for the api to call.
> 2. From the CLI using the upgrade subcommand.


@joshdover could you help out with this? The API calls that we are using are not reflecting what the UI does, or at least that is our observation.

**Comment by mdelapenya** [comment]
@ph @blakerouse what version format does the upgrade subcommand support? I've verified that it does not support hashed snapshots (8.3.0-abcdef-SNAPSHOT), which is used extensively in the test framework. Is this intended?

**Comment by blakerouse** [comment]
@mdelapenya It supports only what would actually be published in the public artifacts repository, which does not include the hash. If the hash snapshot file is placed on disk in a directory and renamed to `8.3.0-SNAPSHOT` then it will upgrade to that file using:

`elastic-agent upgrade 8.3.0-SNAPSHOT --source-uri file://${abs_path_to_dir}`

**Comment by mdelapenya** [comment]
> If the hash snapshot file is placed on disk in a directory and renamed to `8.3.0-SNAPSHOT`

Is there any validation for the file name? I wonder if it's possible to do this:

```
elastic-agent upgrade 8.3.0-SNAPSHOT --source-uri file://tmp/elastic-agent-8.3.0-abcdf-SNAPSHOT-x86_64.rpm
```

**Comment by mdelapenya** [comment]
> > If the hash snapshot file is placed on disk in a directory and renamed to `8.3.0-SNAPSHOT`
> 
> Is there any validation for the file name? I wonder if it's possible to do this:
> 
> ```
> elastic-agent upgrade 8.3.0-SNAPSHOT --source-uri file://tmp/elastic-agent-8.3.0-abcdf-SNAPSHOT-x86_64.rpm
> ```
> 
> I'm looking at https://github.com/elastic/elastic-agent/blob/main/internal/pkg/agent/application/upgrade/step_download.go

@blakerouse I think it's possible. See my PoC https://github.com/elastic/elastic-agent/pull/388

**Comment by mdelapenya** [comment]
When upgrading from 8.3.0-SNAPSHOT to 8.3.0-SNAPSHOT, using the upgrade command, I see this error:

>Error: Failed trigger upgrade of daemon: cannot be upgraded; perform upgrading using Fleet\nFor help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.3/fleet-troubleshooting.html

Same when upgrading from 8.1.2, 8.1.1, 8.1.0 and 7.17-SNAPSHOT to 8.3.0-SNAPSHOT:

>Error: Failed trigger upgrade of daemon: cannot be upgraded; perform upgrading using Fleet\nFor help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/${THE_VERSION}/fleet-troubleshooting.html

**Comment by blakerouse** [comment]
@mdelapenya When the Elastic Agent is controlled by Fleet it must be done by Fleet. In standalone mode the Elastic Agent can be upgraded from the `upgrade` sub-command. So for testing of upgrade best to run them in Standalone mode in this case.

**Comment by joshdover** [comment]
> I'd prefer continuing with the fleet-mode here, as all the code is implemented to enrol the agent it. There was a question for that to @joshdover here [#2354 (comment)](https://github.com/elastic/e2e-testing/pull/2354#issuecomment-1098200304). Waiting on him to determine what differences are using the API calls compared to using the UI.

Apologies, I've been on PTO and still drowning in notifications a bit. I don't see anything obvious jumping out at me as a problem in the curl itself. This API passes down the version number directly without any modifications. Assuming a binary with that version number `8.3.0-657f2d50-SNAPSHOT` really exists for download, then it should be downloaded on the Agent.

That said, the UI uses the version number that includes the snapshot's sha, it will send down `8.3.0-SNAPSHOT` as the version number to grab the latest one. Any reason we can't use that in the test too?

**Comment by mdelapenya** [comment]
> > I'd prefer continuing with the fleet-mode here, as all the code is implemented to enrol the agent it. There was a question for that to @joshdover here [#2354 (comment)](https://github.com/elastic/e2e-testing/pull/2354#issuecomment-1098200304). Waiting on him to determine what differences are using the API calls compared to using the UI.
> 
> Apologies, I've been on PTO and still drowning in notifications a bit. I don't see anything obvious jumping out at me as a problem in the curl itself. This API passes down the version number directly without any modifications. Assuming a binary with that version number `8.3.0-657f2d50-SNAPSHOT` really exists for download, then it should be downloaded on the Agent.
> 
> That said, the UI uses the version number that does not include the snapshot's sha, it will send down `8.3.0-SNAPSHOT` as the version number to grab the latest one. Any reason we can't use that in the test too?

My question is more about how the version is passed: I've verified that it is passed as part of the payload:

```json
{"version":"8.3.0-SNAPSHOT"}
```

but today I've also noticed it's passed as a request header: 
Screenshot 2022-05-03 at 11 06 45

I'm going to add a commit including that header, which was not passed by the tests. Is this something that could affect the result of the upgrade using the APIs? If so, how can the test framework be noticed of those specs/requirements?

**Comment by joshdover** [comment]
> Is this something that could affect the result of the upgrade using the APIs? If so, how can the test framework be noticed of those specs/requirements?

This is a Kibana-wide header that is passed on every request from the UI and it is not read by the upgrade endpoint specifically. It's used to detect when the UI may be out of date from the server after a Kibana upgrade and you shouldn't need to specify it when using the Fleet API.

**Comment by mdelapenya** [comment]
At the test code, we are basically doing (using cURL as an example):

```gherkin
  Given a "<stale-version>" stale agent is deployed to Fleet with "tar" installer
    And certs are installed
    And the "elastic-agent" process is "restarted" on the host
  When agent is upgraded to "latest" version
  Then agent is in "latest" version
```

This is the log for that block:

```
[2022-05-03T10:59:58.886Z] TRAC[2022-05-03T10:59:51Z] Kibana API Query                              body="{\"version\":\"8.3.0-0b6ea9f2-SNAPSHOT\"}" headers="[]" method=POST url="http://3.16.156.84:5601/api/fleet/agents/c1999602-40fe-4c60-90ec-655203b42c3f/upgrade"
[2022-05-03T10:59:58.886Z] TRAC[2022-05-03T10:59:53Z] After step (passed): agent is upgraded to "latest" version 
[2022-05-03T10:59:58.886Z] TRAC[2022-05-03T10:59:53Z] Before step: agent is in "latest" version    
[2022-05-03T10:59:58.886Z] TRAC[2022-05-03T10:59:53Z] Checking if agent is in version 8.3.0-SNAPSHOT. Current version: 8.1.2 
[2022-05-03T10:59:58.886Z] TRAC[2022-05-03T10:59:53Z] Executing command                             args="[/etc/hostname]" command=cat env="map[]"
[2022-05-03T10:59:58.886Z] TRAC[2022-05-03T10:59:53Z] Output                                        output=ip-172-31-45-130
[2022-05-03T10:59:58.886Z] TRAC[2022-05-03T10:59:53Z] Kibana API Query                              body=null headers="[]" method=GET url="http://3.16.156.84:5601/api/fleet/agents"
[2022-05-03T10:59:58.886Z] TRAC[2022-05-03T10:59:53Z] Agent found                                   agent="{c1999602-40fe-4c60-90ec-655203b42c3f e315a3d0-cacf-11ec-9962-d3db60348bfc 2 _oSSiYABHGoyYasFPs27:kG0gO5KeR9iFs17sDvdsrQ {{ip-172-31-45-130 ip-172-31-45-130} {suse SLES(15-SP3) sles} {{8.1.2 false}}} updating}"
[2022-05-03T10:59:58.886Z] WARN[2022-05-03T10:59:53Z] Version mismatch                              elapsedTime=10.002337ms error="version mismatch required '8.3.0-SNAPSHOT' retrieved '8.1.2'" maxTimeout=7m0s retries=1 retrievedVersion=8.1.2 version=8.3.0-SNAPSHOT
[2022-05-03T10:59:58.887Z] TRAC[2022-05-03T10:59:53Z] Kibana API Query                              body=null headers="[]" method=GET url="http://3.16.156.84:5601/api/fleet/agents"
[2022-05-03T10:59:58.887Z] TRAC[2022-05-03T10:59:53Z] Agent found                                   agent="{c1999602-40fe-4c60-90ec-655203b42c3f e315a3d0-cacf-11ec-9962-d3db60348bfc 2 _oSSiYABHGoyYasFPs27:kG0gO5KeR9iFs17sDvdsrQ {{ip-172-31-45-130 ip-172-31-45-130} {suse SLES(15-SP3) sles} {{8.1.2 false}}} updating}"
[2022-05-03T10:59:58.887Z] WARN[2022-05-03T10:59:53Z] Version mismatch                              elapsedTime=573.491505ms error="version mismatch required '8.3.0-SNAPSHOT' retrieved '8.1.2'" maxTimeout=7m0s retries=2 retrievedVersion=8.1.2 version=8.3.0-SNAPSHOT
[2022-05-03T10:59:58.887Z] TRAC[2022-05-03T10:59:55Z] Kibana API Query                              body=null headers="[]" method=GET url="http://3.16.156.84:5601/api/fleet/agents"
[2022-05-03T10:59:58.887Z] TRAC[2022-05-03T10:59:55Z] Agent found                                   agent="{c1999602-40fe-4c60-90ec-655203b42c3f e315a3d0-cacf-11ec-9962-d3db60348bfc 2 _oSSiYABHGoyYasFPs27:kG0gO5KeR9iFs17sDvdsrQ {{ip-172-31-45-130 ip-172-31-45-130} {suse SLES(15-SP3) sles} {{8.1.2 false}}} updating}"
[2022-05-03T10:59:58.887Z] WARN[2022-05-03T10:59:55Z] Version mismatch                              elapsedTime=2.025022353s error="version mismatch required '8.3.0-SNAPSHOT' retrieved '8.1.2'" maxTimeout=7m0s retries=3 retrievedVersion=8.1.2 version=8.3.0-SNAPSHOT
[2022-05-03T10:59:58.887Z] TRAC[2022-05-03T10:59:57Z] Kibana API Query                              body=null headers="[]" method=GET url="http://3.16.156.84:5601/api/fleet/agents"
[2022-05-03T10:59:58.887Z] TRAC[2022-05-03T10:59:57Z] Agent found                                   agent="{c1999602-40fe-4c60-90ec-655203b42c3f e315a3d0-cacf-11ec-9962-d3db60348bfc 2 _oSSiYABHGoyYasFPs27:kG0gO5KeR9iFs17sDvdsrQ {{ip-172-31-45-130 ip-172-31-45-130} {suse SLES(15-SP3) sles} {{8.1.2 false}}} online}"
[2022-05-03T10:59:58.888Z] WARN[2022-05-03T10:59:57Z] Version mismatch                              elapsedTime=4.367357992s error="version mismatch required '8.3.0-SNAPSHOT' retrieved '8.1.2'" maxTimeout=7m0s retries=4 retrievedVersion=8.1.2 version=8.3.0-SNAPSHOT
[2022-05-03T11:00:01.427Z] TRAC[2022-05-03T11:00:01Z] Kibana API Query                              body=null headers="[]" method=GET url="http://3.16.156.84:5601/api/fleet/agents"
[2022-05-03T11:00:01.427Z] TRAC[2022-05-03T11:00:01Z] Agent found                                   agent="{c1999602-40fe-4c60-90ec-655203b42c3f e315a3d0-cacf-11ec-9962-d3db60348bfc 2 _oSSiYABHGoyYasFPs27:kG0gO5KeR9iFs17sDvdsrQ {{ip-172-31-45-130 ip-172-31-45-130} {suse SLES(15-SP3) sles} {{8.1.2 false}}} online}"
[2022-05-03T11:00:01.427Z] WARN[2022-05-03T11:00:01Z] Version mismatch                              elapsedTime=8.133198805s error="version mismatch required '8.3.0-SNAPSHOT' retrieved '8.1.2'" maxTimeout=7m0s retries=5 retrievedVersion=8.1.2 version=8.3.0-SNAPSHOT
[2022-05-03T11:00:06.716Z] TRAC[2022-05-03T11:00:05Z] Kibana API Query                              body=null headers="[]" method=GET url="http://3.16.156.84:5601/api/fleet/agents"
[2022-05-03T11:00:06.716Z] TRAC[2022-05-03T11:00:05Z] Agent found                                   agent="{c1999602-40fe-4c60-90ec-655203b42c3f e315a3d0-cacf-11ec-9962-d3db60348bfc 2 _oSSiYABHGoyYasFPs27:kG0gO5KeR9iFs17sDvdsrQ {{ip-172-31-45-130 ip-172-31-45-130} {suse SLES(15-SP3) sles} {{8.1.2 false}}} error}"
```

From the above logs, where we are doing the following:

1. we send a POST request to Fleet in /api/fleet/agents/c1999602-40fe-4c60-90ec-655203b42c3f/upgrade , using `{"version":"8.3.0-0b6ea9f2-SNAPSHOT"}` as payload.
2. after the first GET to /api/fleet/agents, filtering by agentID (`c1999602-40fe-4c60-90ec-655203b42c3f`), we get the agent is in the `updating` state:
>{c1999602-40fe-4c60-90ec-655203b42c3f e315a3d0-cacf-11ec-9962-d3db60348bfc 2 _oSSiYABHGoyYasFPs27:kG0gO5KeR9iFs17sDvdsrQ {{ip-172-31-45-130 ip-172-31-45-130} {suse SLES(15-SP3) sles} {{8.1.2 false}}} updating}

3. in the same millisecond there are 3 calls to that API, retrieving the agent in the `updating` status.
4. in the same millisecond, after the forth call, the API retrieves the agent is in the `online` status:
>{c1999602-40fe-4c60-90ec-655203b42c3f e315a3d0-cacf-11ec-9962-d3db60348bfc 2 _oSSiYABHGoyYasFPs27:kG0gO5KeR9iFs17sDvdsrQ {{ip-172-31-45-130 ip-172-31-45-130} {suse SLES(15-SP3) sles} {{8.1.2 false}}} online}
5. 2'5 seconds after that, a new GET request retrieves the agent is in the `online` status but in the wrong version.
6. 5 seconds after that, the next GET request, retrieves the agent in the `error` status and in the wrong version:
>{c1999602-40fe-4c60-90ec-655203b42c3f e315a3d0-cacf-11ec-9962-d3db60348bfc 2 _oSSiYABHGoyYasFPs27:kG0gO5KeR9iFs17sDvdsrQ {{ip-172-31-45-130 ip-172-31-45-130} {suse SLES(15-SP3) sles} {{8.1.2 false}}} error}

For some reason, although the upgrade command is submitted to Fleet successfully (the status changed!) the status goes to error after a while.

**Review by cachedout — APPROVED** [review]

**Comment by mdelapenya** [comment]
After adding the `kbn-version` header using a hashed SNAPSHOT version (ex: 8.3.0-abcdef-SNAPSHOT), the upgrade fails with:

> [2022-05-05T08:25:04.378Z] ERRO[2022-05-05T08:24:55Z] Could not upgrade agent                       body="{'statusCode':400,'error':'Bad Request','message':'Browser client is out of date, please refresh the page (\\'kbn-version\\' header was \\'8.3.0-a1c5cfff-SNAPSHOT\\' but should be \\'8.3.0-SNAPSHOT\\')','attributes':{'expected':'8.3.0-SNAPSHOT','got':'8.3.0-a1c5cfff-SNAPSHOT'}}" error="<nil>" statusCode=400

File: e2e/_suites/fleet/fleet.go [hunk 2]
```diff
@@ -384,45 +362,70 @@ func (fts *FleetTestSuite) installCerts() error {
...
 	case "latest":
 		desiredVersion = common.ElasticAgentVersion
-	default:
-		desiredVersion = common.ElasticAgentVersion
 	}
+	log.Tracef("Desired version is %s. Current version: %s", desiredVersion, fts.Version)
 
 	agentService := deploy.NewServiceRequest(common.ElasticAgentServiceName)
+
+	/*
```
Comment (mdelapenya) [inline e2e/_suites/fleet/fleet.go:390]: Keeping the upgrade-using-command code in a comment, but I will remove it before merging the PR

**Comment by mdelapenya** [comment]
@jlind23 @ph this is working now!

image

The plan would include:
- define in what platforms should the upgrade scenarios be run. It's running on Debian 10 AMD, Debian 10 ARM and SLES15. See https://github.com/elastic/e2e-testing/pull/2354/files#diff-b87fe24438639288bccba6a647eac526dec8e1755dcf1fc7959434a3ba95404aR46
- define in what schedule to run these scenarios: nightly? per PR? I noticed they do not add that much overhead to the build, as there are other parallel branches taking longer.
- merge into main
- manually backport to the different maintenance branches updating the Gherkin examples to match the stale-version for each maintenance branch. Current stale versions: https://github.com/elastic/e2e-testing/pull/2354/files#diff-3746d5ea011663df7d08a904fbe1d56d2ed37824faaa9bbd7669a749f7657d79R12-R19
- check if the scenarios should be included for Beats changes. It's possible to define "profiles" that can be used by consumers of the e2e tests to select what scenarios and suites should be run for that consumer.

**Comment by jlind23** [comment]
@mdelapenya as discussed today, those are the Operating systems that must be used on every PRs:
- Windows 11
- MacOs 12
- Ubuntu 20 (because our docker will be based on this)
- RHEL 8 (64 bit ARM)

Then can we run all scenarios, on all platform, twice a day?


cc @ph @joshdover

**Comment by nchaulet** [comment]
> After adding the kbn-version header using a hashed SNAPSHOT version (ex: 8.3.0-abcdef-SNAPSHOT), the upgrade fails with:

I think the `kbn-version` header purpose is to be used in Kibana client side code, for other clients we probably want to use `kbn-xsrf` header https://www.elastic.co/guide/en/kibana/current/api.html#api-request-headers

**Comment by mdelapenya** [comment]
> > After adding the kbn-version header using a hashed SNAPSHOT version (ex: 8.3.0-abcdef-SNAPSHOT), the upgrade fails with:
> 
> I think the `kbn-version` header purpose is to be used in Kibana client side code, for other clients we probably want to use `kbn-xsrf` header https://www.elastic.co/guide/en/kibana/current/api.html#api-request-headers

Yes, we added although it's useless for this purpose. Will remove it from the code

**Comment by mdelapenya** [comment]
I'm merging this one. Will update the backport to match the right stale versions for each maintenance branch, and will create an issue to define in which OSs and build schedule to run these tests.

Thanks!

**Pull request merged**